### DISCUSSION TITLE
Alerts & Reports with extra state

### DISCUSSION THREAD

**Discussion by villebro** [apache/superset#24843] [Ideas]
Currently Alerts & Reports only supports rendering the default view of charts and dashboards. Especially for dashboards, this can be limiting, as it may be necessary to render a dashboard with certain predefined native filters, or a specific active tab. Fortunately we already have permalinks that can store information. For instance, for dashboards, the permalinks record the following info:
- Filter state
- The active tab
- Anchor to active chart
- URL params

To be able to schedule reports that reference a dashboard with specific filter state (and/or URL params, active tabs/charts etc), we could add an optional field to the Alerts/Reports modal that would accept a permalink:
<img width="1187" alt="image" src="https://github.com/apache/superset/assets/33317356/69e3e8b2-6822-4345-8393-0865629c7240">

After the permalink is parsed, the dashboard/chart value would be automatically populated, with the permalink state captured in a separate field (if a chart permalink was used, the chart would be selected):
<img width="1188" alt="image" src="https://github.com/apache/superset/assets/33317356/984d4631-e6c0-42b9-8fdb-a16be977d55a">

To change the permalink, one could either select the Permalink option again and post a full permalink (which would select the dashboard/chart radio button and pre-populate the chart/dashboard value + permalink), or just update the permalink id in the respective text field. After this, the dashboard would be rendered with the appropriate filters + tabs pre-selected:
<img width="1403" alt="image" src="https://github.com/apache/superset/assets/33317356/eb60c29a-0850-45ac-b93c-ac73375dcb27">

Thoughts? Ping @betodealmeida @rusackas @michael-s-molina @john-bodley @kasiazjc @nytai @anamitraadhikari

**Comment by michael-s-molina** [comment]
Thank you the discussion @villebro. I think this feature is very useful and will definitely increase the configurability of Alerts & Reports. Regarding the proposed UI, I think the Permalink should not be in the same information level as Dashboards and Charts given that is applicable to both. Maybe it could be a field that you can fill after you select Dashboard or Chart and it will auto populate the select.

Another comment is to try to avoid URL parsing and just save/display the full permalink URL. If the user wishes to update a permalink they will probably copy/paste the value again.

**Reply by villebro** [reply]
Thanks for the feedback @michael-s-molina . The proposed UI changes sound fine to me. So to summarize, it would work as follows for a typical dashboard permalink case:
1. User chooses "Dashboard"
2. Instead of selecting a dashboard from the dropdown, they paste a permalink, after which the dashboard is automatically populated based on the state in the permalink
3. After saving and reopening the alert/report, the user will see the dashboard preselected in the dropdown, with the full permalink URL

Correct? @kasiazjc do you have any thoughts on this?

**Reply by kasiazjc** [reply]
I think it's getting crowded with another radio button, so I would suggest switching from radio buttons to single select field. Then it would be more clear what the child filed is referring to and it could change dynamically. Adding a screenshot below. 

I wonder if in case of selecting permalink, the dashboard select dropdown should disabled (but populated). Because what if someone changes dashboard, but not the permalink? Thoughts? 

<img width="407" alt="image" src="https://github.com/apache/superset/assets/36897697/4e3b73c0-ae00-447b-8750-12c8a29f59a1">
<img width="413" alt="image" src="https://github.com/apache/superset/assets/36897697/5b71779d-da45-487d-9c87-f9519c9e018c">

**Reply by michael-s-molina** [reply]
If we treat Permalink at the same level of Dashboard/Chart using a select, then I suggest showing the dashboard name as text instead of disabled select as it will be more readable. We just need to make sure the text correctly identifies that the asset is a dashboard or chart so I would keep the header (without the required icon). Something like:

Dashboard:
My beautiful dashboard

Chart:
My beautiful chart

And instead of `Permalink state`, which is a hard concept for users to understand, I would have `Permalink URL` or just `Permalink` which can be easily copied/pasted from the `Copy permalink to clipboard` option.

**Reply by kasiazjc** [reply]
That makes sense! Thank you. I think we can add a tooltip to permalink too and explain where to find it/what it is 
(also * are my fault and random as I was trying something out here, sorry for confusion!)

**Comment by kasiazjc** [comment]
Question - what would happen if someone deletes a filter that is stored in the permalink or something like this? Would screenshot still be generated, generated blank or would user get some kind of notification?

**Reply by michael-s-molina** [reply]
Great question. If a filter is deleted, the permalink is not able to restore it but all the other filters are preserved. So if I have a permalink for City = Johns Creek and State = Georgia and someone deletes the City filter, the dashboard will be rendered without the City filter but with State = Georgia.

**Reply by kasiazjc** [reply]
That makes sense and (for now) I think it's an expected behavior. Mentioned it to Ville too, but if possible in the future (def not now), when deleting a filter which is used in reports we could add a warning that by deleting it, report/screenshot will be altered.

**Comment by mattitoo** [comment]
I really like this idea

**Comment by mattitoo** [comment]
Actually, someone already developed the permalink feature: https://apache-superset.slack.com/archives/CCKHMGRRB/p1692794008128599

**Reply by vivek-kandhvar** [reply]
seems like the archive is gone!

**Comment by vivek-kandhvar** [comment]
Though this feature can be very handy, it's not really user intuitive.

One dashboard or chart can have multiple permalinks at any point of time, and the User will not really know what filter the permalink actually holds. Also, user has to first go to a dashboard or chart, select all required filters, generate a permalink, and then come back to Alerts and Reports, and then create a report, and paste this link. That is way too many steps.

I would prefer to enhance permalink feature to add a user defined name to it, add a name column in the table, and then show the associated filters in a drop-down to be applied, when creating a report.

**Reply by jitendra-dangi** [reply]
Any update on this proposal?

**Comment by villebro** [comment]
> Though this feature can be very handy, it's not really user intuitive.
> 
> One dashboard or chart can have multiple permalinks at any point of time, and the User will not really know what filter the permalink actually holds. Also, user has to first go to a dashboard or chart, select all required filters, generate a permalink, and then come back to Alerts and Reports, and then create a report, and paste this link. That is way too many steps.
> 
> I would prefer to enhance permalink feature to add a user defined name to it, add a name column in the table, and then show the associated filters in a drop-down to be applied, when creating a report.

@vivek-kandhvar while this could be useful, the changes you propose should IMO be considered a separate feature, and one that probably warrants a SIP. My proposal is meant to simply pull in the functionality that's already supported by permalinks, and unlock rendering different views of charts/dashboards that's currently not possible. We can later consider expanding on this feature if the UX is found to be too poor by users (adding new state to permalinks and/or making the state editable in the report modal etc).
